main.tf

```diff
@@ -56,52 +100,198 @@ resource "aws_launch_configuration" "this" {
   }
 }
 
-####################
-# Autoscaling group
-####################
-resource "aws_autoscaling_group" "this" {
-  count = var.create_asg && false == var.create_asg_with_initial_lifecycle_hook ? 1 : 0
-
-  name_prefix = "${join(
-    "-",
-    compact(
-      [
-        coalesce(var.asg_name, var.name),
-        var.recreate_asg_when_lc_changes ? element(concat(random_pet.asg_name.*.id, [""]), 0) : "",
-      ],
-    ),
-  )}-"
-  launch_configuration = var.create_lc ? element(concat(aws_launch_configuration.this.*.name, [""]), 0) : var.launch_configuration
-  vpc_zone_identifier  = var.vpc_zone_identifier
-  max_size             = var.max_size
-  min_size             = var.min_size
-  desired_capacity     = var.desired_capacity
+################################################################################
+# Launch template
+################################################################################
 
-  load_balancers            = var.load_balancers
-  health_check_grace_period = var.health_check_grace_period
-  health_check_type         = var.health_check_type
+resource "aws_launch_template" "this" {
+  count = var.create_lt ? 1 : 0
 
-  min_elb_capacity          = var.min_elb_capacity
-  wait_for_elb_capacity     = var.wait_for_elb_capacity
-  target_group_arns         = var.target_group_arns
-  default_cooldown          = var.default_cooldown
-  force_delete              = var.force_delete
-  termination_policies      = var.termination_policies
-  suspended_processes       = var.suspended_processes
-  placement_group           = var.placement_group
-  enabled_metrics           = var.enabled_metrics
-  metrics_granularity       = var.metrics_granularity
-  wait_for_capacity_timeout = var.wait_for_capacity_timeout
-  protect_from_scale_in     = var.protect_from_scale_in
-  service_linked_role_arn   = var.service_linked_role_arn
-  max_instance_lifetime     = var.max_instance_lifetime
+  name        = var.lt_use_name_prefix ? null : local.lt_name
+  name_prefix = var.lt_use_name_prefix ? "${local.lt_name}-" : null
+  description = var.description
+
+  ebs_optimized = var.ebs_optimized
+  image_id      = var.image_id
+  instance_type = var.instance_type
+  key_name      = var.key_name
+  user_data     = var.user_data_base64
```

Comment:
`user_data_base64` or `user_data`?

Reply:
with launch template, they only accept base64 encoded userdata (launch config has two separate attributes for plain text or base64 encoded)
<img width="661" alt="image" src="https://user-images.githubusercontent.com/10913471/113707535-0bb7c180-96ae-11eb-90d2-ab847adc8ccd.png">

Reply:
Good to know!